File: verion1/fetch_data.php
<?php
header('Content-Type: application/json');
include "connect.php";

// Function to ping an IP address
function ping($ip) {
    // Execute the ping command and capture both output and return status
    $pingResult = shell_exec(sprintf('ping -n 1 -w 1000 %s 2>&1', escapeshellarg($ip))); // Redirect errors to output
    // error_log("Ping command executed for IP $ip. Result: $pingResult"); // Log the full result for debugging

    // Check if the output contains 'TTL' (indicating a successful ping)
    if (strpos($pingResult, 'TTL') !== false) {
        return true;
    }

    // Log if the ping failed
    // error_log("Ping failed for IP $ip. Check network or firewall settings.");
    return false;
}

// Fetch data from `bag_info`
$sql = "SELECT id, name, active_status, ivbag_level, backflow, room, ipaddress, macaddress, ivbag_type FROM bag_info";
$result = $conn->query($sql);

$data = [];
if ($result->num_rows > 0) {
    while ($row = $result->fetch_assoc()) {
        $row['ip_active'] = ping($row['ipaddress']) ? true : false; // Add IP active status
        $data[] = $row;
    }
}
if (isset($_GET['fetch_updates'])) {
    $sql = "SELECT id, name, active_status, ivbag_level, backflow, room, ipaddress, macaddress, ivbag_type FROM bag_info";
    $result = $conn->query($sql);

    $data = [];
    if ($result->num_rows > 0) {
        while ($row = $result->fetch_assoc()) {
            $row['ip_active'] = ping($row['ipaddress']) ? true : false; // Add IP active status
            $data[] = $row;
        }
    }
    echo json_encode($data);
    $conn->close();
    exit;
}
// Return JSON response
echo json_encode($data);
$conn->close();

File: verion1/fetch_data.py
from flask import Flask, jsonify, request
import pymysql
from ping3 import ping

app = Flask(__name__)

# Database connection details
DB_CONFIG = {
    "host": "localhost",
    "user": "root",
    "password": "",
    "database": "ivbag_data"
}

# Function to ping an IP address
def is_ip_active(ip):
    try:
        response = ping(ip, timeout=1)  # Ping with a 1-second timeout
        return response is not None
    except Exception as e:
        app.logger.error(f"Error pinging IP {ip}: {e}")
        return False

@app.route('/fetch_data', methods=['GET'])
def fetch_data():
    fetch_updates = request.args.get('fetch_updates', default=False, type=bool)
    connection = pymysql.connect(**DB_CONFIG)
    cursor = connection.cursor(pymysql.cursors.DictCursor)

    try:
        query = "SELECT id, name, active_status, ivbag_level, backflow, room, ipaddress FROM bag_info"
        cursor.execute(query)
        rows = cursor.fetchall()

        data = []
        for row in rows:
            row['ip_active'] = is_ip_active(row['ipaddress'])
            data.append(row)

        return jsonify(data)
    except Exception as e:
        app.logger.error(f"Error fetching data: {e}")
        return jsonify({"error": "Failed to fetch data"}), 500
    finally:
        cursor.close()
        connection.close()

if __name__ == '__main__':
    app.run(debug=True, host='0.0.0.0', port=5000)
